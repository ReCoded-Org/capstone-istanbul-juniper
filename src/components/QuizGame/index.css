.quizWrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
  background: #b4d0c4;
  font-size: 20px;
}

.quizWrapper__score {
  color: #fff;
  font-size: 2rem;
  margin: 0;
}

.quizWrapper__header {
  font-family: "Fascinate" Inline;
  background-image: linear-gradient(180deg, #fff, #87f1ff);
  font-weight: 400;
  background-size: 100%;
  background-clip: text;
  filter: drop-shadow(2px 2px #0085a3);
  font-size: 70px;
  text-align: center;
  margin: 20px;
}

.quizWrapper__startGame,
.quizWrapper__nextButton {
  cursor: pointer;
  background-color: #f8ac4c;
  border: 2px solid #d38558;
  box-shadow: 0px 5px 10px rgba(0, 0, 0, 0.25);
  border-radius: 10px;
  height: 40px;
  margin: 20px 0;
  padding: 0 40px;
  outline: none;
  background: #b4d0c4;
}

.quizWrapper__startGame {
  max-width: 200px;
  background-color: #f8ac4c;
}

.quizWrapper__nextButton {
  background-color: #f8ac4c;
}
