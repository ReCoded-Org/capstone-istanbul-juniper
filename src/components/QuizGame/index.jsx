import React, { useState } from "react";
import { fetchQuizQuestions } from "../../components/QuizGameApi";
import QuestionCard from "../../components/QuizGameQuestions";
import "./index.css";

const DIFFICULTY = {
  EASY: "easy",
  MEDIUM: "medium",
  HARD: "hard",
};

const TOTAL_QUESTIONS = 10;

const QuizGame = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [questions, setQuestions] = useState([]);
  const [currentQuestionNumber, setCurrentQuestionNumber] = useState(0);
  const [userAnswers, setUserAnswers] = useState([]);
  const [score, setScore] = useState(0);
  const [isGameStart, setIsGameStart] = useState(false);

  const startQuiz = async () => {
    setIsLoading(true);
    setIsGameStart(true);
    const newQuestions = await fetchQuizQuestions(
      TOTAL_QUESTIONS,
      DIFFICULTY.EASY
    );
    setQuestions(newQuestions);
    setScore(0);
    setUserAnswers([]);
    setCurrentQuestionNumber(0);
    setIsLoading(false);
  };

  const checkAnswer = (e) => {
    if (isGameStart) {
      const userChosenAnswer = e.currentTarget.value;
      const solutionAnswer = questions[currentQuestionNumber].correct_answer;
      const isCorrect = solutionAnswer === userChosenAnswer;
      if (isCorrect) {
        setScore((prev) => prev + 1);
        const answerObject = {
          question: questions[currentQuestionNumber].question,
          userChosenAnswer: e.currentTarget.value,
          isCorrect: solutionAnswer === userChosenAnswer,
          correctAnswer: questions[currentQuestionNumber].correct_answer,
        };
        setUserAnswers((prev) => [...prev, answerObject]);
      }
    }
  };

  const nextQuestion = () => {
    const nextQNum = currentQuestionNumber + 1;
    if (nextQNum >= TOTAL_QUESTIONS) {
      // the start button will appear again after the user finish ten questions and of course to start the game.
      setIsGameStart(true);
    } else {
      setCurrentQuestionNumber(nextQNum);
    }
  };

  const shouldStartGameAgain =
    !isGameStart || userAnswers.length === TOTAL_QUESTIONS;
  const isGameReady = isGameStart && !isLoading;
  const isUserOnLastQuestion =
    userAnswers.length === currentQuestionNumber + 1 &&
    currentQuestionNumber !== TOTAL_QUESTIONS - 1;
  const isThereAnotherQuestion = isGameReady && isUserOnLastQuestion;

  return (
    <>
      <div className="quizWrapper">
        <h1 className="quizWrapper__header">Quiz Game</h1>
        {shouldStartGameAgain && (
          <button className="quizWrapper__startGame" onClick={startQuiz}>
            Start
          </button>
        )}
        {isGameStart && <p className="quizWrapper__score">Score: {score}</p>}
        {isLoading && <p>Loading Questions...</p>}
        {isGameReady && (
          <QuestionCard
            questionNum={currentQuestionNumber + 1}
            totalQuestions={TOTAL_QUESTIONS}
            question={questions[currentQuestionNumber].question}
            answerChoices={questions[currentQuestionNumber].answers}
            handleClick={checkAnswer}
          />
        )}
        {isThereAnotherQuestion && (
          <button className="quizWrapper__nextButton" onClick={nextQuestion}>
            Next question
          </button>
        )}
      </div>
    </>
  );
};
export default QuizGame;
