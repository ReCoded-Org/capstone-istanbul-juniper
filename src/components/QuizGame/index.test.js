import React from "react";
import QuizGame from "./index.jsx";
import renderer from "react-test-renderer";

it("Quiz Game snapshot test", () => {
  const tree = renderer.create(<QuizGame />).toJSON();
  expect(tree).toMatchSnapshot();
});
