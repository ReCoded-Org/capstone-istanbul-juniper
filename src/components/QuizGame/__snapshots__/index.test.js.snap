// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Game snapshot test 1`] = `
<div
  className="quizWrapper"
>
  <h1
    className="quizWrapper__header"
  >
    Quiz Game
  </h1>
  <button
    className="quizWrapper__startGame"
    onClick={[Function]}
  >
    Start
  </button>
</div>
`;
